### ISSUE TITLE
Allow Parallel Recording of RenderPasses

### ISSUE THREAD

**Issue by cwfitzgerald** [gfx-rs/wgpu#1104]
**Is your feature request related to a problem? Please describe.**

Currently, the best way to multithreaded the command recording of wgpu applications is to spread the job of recording onto multiple threads. However, due to a singular write lock we take on the command buffer list [here](https://github.com/gfx-rs/wgpu/blob/master/wgpu-core/src/command/render.rs#L522), it enforces that only one RenderPass will be running at once.

**Describe the solution you'd like**

Demote this write lock to a read lock. The simplest way is to make the registry hold `Registry<RwLock<CommandBuffer<B>>, CommandBufferId, F>`. There may be better ideas.

**Describe alternatives you've considered**

None. This is a pretty vital feature and one of the big appeals of modern apis, so we want to make sure this works well.

**Additional context**

This came up when someone was trying to improve encoding time of a program with tens of thousands of direct draw calls. Parallelizing the job made it perform no better.

**Comment by kvark** [comment]
I have another idea that would be simpler to roll in. During the recording itself, we don't need a mutable command buffer at all. We need to read the base resource tracker on it, we need to read the thread ID, maybe a few other things. But the point is - we only need to read it for the duration of the actual pass encoding (which is most time consuming).

The only mutable actions we need is inserting the barriers and modifying the trackers, as well as appending the newly recorded command buffer to the end. There isn't much to [do there](https://github.com/gfx-rs/wgpu/blob/2287ae3f8a7246e8b47b73a1de6fe2410c42b8c8/wgpu-core/src/command/render.rs#L1768-L1778).

So maybe we can start by refactoring `command_encoder_run_render_pass_impl` to do the read-only access of the command buffer storage while it's recording, and then drop the guard and re-acquire it for mutable access for a very short time? I could see how this would let multiple threads do more things at once, but of course in other cases it will not help at all (i.e. we are finished recording on one thread, trying to re-lock the storage, but have to wait for another thread to also finish recording).

**Comment by cwfitzgerald** [comment]
So I ended up hacking together a proof of concept of what I proposed to see if it was going to work (and to test for other bottlenecks): https://github.com/cwfitzgerald/wgpu/tree/thread-fix. Code is a bit of a mess, but works as expected, and gives a 4x speedup with multithreaded recording on a 4C/8T machine. I don't know if this is a viable way forward from being not just a hack

I also made a benchmark, I modified the cube example to record 10,000 different instances of the cube over an arbitrary amount of threads using rayon. https://github.com/cwfitzgerald/wgpu-rs/tree/cube-stress We can try the way you described too and see how well it ends up doing.
